src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateSignature.cs

```diff
@@ -0,0 +1,314 @@
+                    jsonWebToken.EncodedToken,
+                    0,
+                    jsonWebToken.Dot2,
+                    Encoding.UTF8,
+                    jsonWebToken.EncodedToken,
+                    jsonWebToken.Dot2,
+                    signatureProvider,
+                    ValidateSignature);
+
+                if (valid)
+                    return new SignatureValidationResult();
+                else
+                    return new SignatureValidationResult(
+                        ValidationFailureType.SignatureValidationFailed,
+                        new ExceptionDetail(
+                            new MessageDetail(TokenLogMessages.IDX10504),
+                            typeof(SecurityTokenInvalidSignatureException),
+                            new StackFrame()));
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
+            catch (Exception ex)
+#pragma warning restore CA1031 // Do not catch general exception types
+            {
+                return new SignatureValidationResult(
+                    ValidationFailureType.SignatureValidationFailed,
+                    new ExceptionDetail(
+                        new MessageDetail(TokenLogMessages.IDX10504, ex.ToString()),
+                        ex.GetType(),
+                        new StackFrame(),
+                        ex));
+            }
+            finally
+            {
+                cryptoProviderFactory.ReleaseSignatureProvider(signatureProvider);
+            }
+        }
+
+        private static ExceptionDetail GetSignatureValidationFailureExceptionDetails(
+            JsonWebToken jwtToken,
+            ValidationParameters validationParameters,
+            BaseConfiguration? configuration,
+            StringBuilder? exceptionStrings,
+            StringBuilder? keysAttempted,
+            bool kidExists,
+            bool kidMatched)
+        {
+            // Get information on where keys used during token validation came from for debugging purposes.
+            var keysInTokenValidationParameters = validationParameters.IssuerSigningKeys;
+            var keysInConfiguration = configuration?.SigningKeys;
+            var numKeysInTokenValidationParameters = keysInTokenValidationParameters.Count;
+            var numKeysInConfiguration = keysInConfiguration?.Count ?? 0;
+
+            if (kidExists && kidMatched)
+            {
+                JsonWebToken localJwtToken = jwtToken; // avoid closure on non-exceptional path
+                var isKidInTVP = keysInTokenValidationParameters.Any(x => x.KeyId.Equals(localJwtToken.Kid));
+                var keyLocation = isKidInTVP ? "TokenValidationParameters" : "Configuration";
+                return new ExceptionDetail(
+                    new MessageDetail(
+                        TokenLogMessages.IDX10511,
+                        LogHelper.MarkAsNonPII(keysAttempted?.ToString() ?? ""),
```

Comment:
do we need a check if logging is enabled first?

Reply:
This is basically the message attached to the exception when generated. This will not log anything.
I thought whether we should have this information given it was information before, but now that we are returning an error, we need to tell something back to the user and that seemed to fit best.